elasticdl/python/worker/worker_pytorch.py

```diff
@@ -0,0 +1,498 @@
+            dense_params, uninit_ps = self._ps_client.pull_dense_parameters(
+                [i for i in range(self._ps_client.ps_num)],
+                self._model_versions_from_ps,
+            )
+
+            # 2. Worker pushes local dense parameters to these PS instances
+            # to initialize their partition of parameters.
+            if len(uninit_ps) > 0:
+                for ps_id in uninit_ps:
+                    # push variable to ps for initialization
+                    parameters = [
+                        Tensor(name, self._non_embed_vars[name].numpy(), None)
+                        for name in self._ps_client.ps_to_parameter[ps_id]
+                    ]
+                    self._ps_client.push_dense_parameters(
+                        parameters, ps_id, self._model_versions_from_ps[ps_id]
+                    )
+
+                ps_params, uninit = self._ps_client.pull_dense_parameters(
+                    uninit_ps, self._model_versions_from_ps
+                )
+                if len(uninit) > 0:
+                    # TODO: support PS fault-tolerance
+                    raise RuntimeError("PS initialization failed")
+                dense_params.update(ps_params)
+
+            # 3. Assign parameters to local model
+            for k, v in dense_params.items():
+                self._non_embed_vars[k] = torch.from_numpy(v)
+            self._model_version = max(self._model_versions_from_ps)
+        self._timing.end_record_time("get_model")
+
+    def report_gradient_to_ps(self, gradients):
+        """
+        report gradient in numpy
+        report learning_rate about PyTorch model optimizer
+        TODO: PS is based on TensorFlow, can not report learning_rate with PyTorch callback
+        """
+        self._timing.start_record_time("report_gradient")
+        grads = []
+        for i, v in enumerate(self._non_embed_vars.items()):
+            grad = Tensor(v[0], gradients[i].numpy(), None)
+            grads.append(grad)
+        edl_grads = []
+        learning_rate = 0.001
+        # learning_rate = self._model.optimizer.param_groups[0]["lr"]
+        accepted, max_version = self._ps_client.push_gradients(
+            grads, edl_grads, learning_rate, self._model_versions_from_ps,
+        )
+        self._timing.end_record_time("report_gradient")
+        return accepted, max_version
+
+    def report_gradient(self, grads):
+        if (
+                self._distribution_strategy
+                == DistributionStrategy.PARAMETER_SERVER
+        ):
+            return self.report_gradient_to_ps(grads)
+        else:
+            raise RuntimeError(
+                "Only support Allreduce and ParameterServer "
+                "distribution strategy"
+            )
+
+    def _run_model_call_before_training(self, features):
+        """
+        before training: Create variables and report to ps if not created.
+        """
+        _ = self._model.forward(features)
```

Comment:
No need to call `forward` here for pytorch.

Reply:
maybe like this: `self._model(features)`?

Reply:
PyTorch nn.Module model is similar to Keras functional model. When we instance a model, all variables are created and initialized. 
For Keras subclass model, variables are lazy-created. Thus, a forward call is needed for subclass model.

Reply:
PyTorch may need it only when we need to substitute embedding with ElasticDL embedding.

Reply:
Thanks, I figured out the problem.